### ISSUE TITLE
The loss function seems to be error

### ISSUE THREAD

**Issue by Woolseyyy** [ronghuaiyang/arcface-pytorch#29]
The focal loss function seems to be differernt from the theory. 
The theory describes the loss as: 
-alpha * (1-y')^gamma * log(y')  - (1-alpha) * y'^gamma * log(1-y') 
Besides, the arc metric also confuses me...ArcFace is modified based on softmax but there is nothing similiar to softmax in the metric code.
Maybe I misunderstand the theory. I will be grateful if you point my errors out.

**Comment by Woolseyyy** [comment]
About the seconde question, I have found the answer. Torch.nn.CrossEntropyLoss() combines nn.LogSoftmax() and nn.NLLLoss() in one single class.

**Comment by CodeMonkZy** [comment]
> About the seconde question, I have found the answer. Torch.nn.CrossEntropyLoss() combines nn.LogSoftmax() and nn.NLLLoss() in one single class.

have you try sphereloss in metric.py？i can`t make loss decrease

**Comment by buaazyz** [comment]
have you solve this problem? Is the 'ArcMarginProduct' of metric.py the same as the theory? Does it work? thank you.

**Comment by ReverseSystem001** [comment]
> > About the seconde question, I have found the answer. Torch.nn.CrossEntropyLoss() combines nn.LogSoftmax() and nn.NLLLoss() in one single class.
> 
> have you try sphereloss in metric.py？i can`t make loss decrease

my training loss is not decline too, how do you resolve this problem???

**Comment by LucasWangZH** [comment]
I think the focal loss here is wrong.

**Comment by wenzhengzeng** [comment]
> About the seconde question, I have found the answer. Torch.nn.CrossEntropyLoss() combines nn.LogSoftmax() and nn.NLLLoss() in one single class.

So is the implementation here of  ArcFace loss right?
